lean4-infoview/src/infoview/rpcInterface.ts

```diff
@@ -183,3 +184,48 @@ export async function getGoToLocation(rs: RpcSessions, pos: DocumentPosition, ki
     const args: GetGoToLocationParams = { kind, info };
     return rs.call<LocationLink[]>(pos, 'Lean.Widget.getGoToLocation', args)
 }
+
+export interface UserWidget {
+    id: string;
+    /** Pretty user name. */
+    name: string;
+    /** A hash (provided by Lean) of the widgetSource's sourcetext.
+     * This is used to look up the WidgetSource object.
+     */
+    javascriptHash: string;
+}
+
+/** Represents an instance of a user widget that can be rendered.
+ * This is used as the input to the `UserWidget` component.
+ */
+export interface UserWidgetInstance extends UserWidget {
+    /** JSON object to be passed as props to the component */
+    props : any
+    range?: Range
+}
+
+/** The response type for the RPC call `Widget_getWidgets`. */
+export interface UserWidgets {
+    widgets : UserWidgetInstance[]
+}
+
+/** Given a position, returns all of the user-widgets on the infotree at this position. */
+export function Widget_getWidgets(rs: RpcSessions, pos: DocumentPosition): Promise<UserWidgets | undefined> {
```

Comment:
I started on the merge, but noticed the following thing: in our RPC calls, sending `DocumentPosition` actually makes no sense because we always use the URI at which the RPC call was made (so now, the `RpcSessionAtPos` URI). `getInteractiveDiagnostics` just takes in a line range, whereas `getInteractiveGoals` takes the whole `TextDocumentPositionParams` but the the handler just ignores it. Therefore I think we should change the widget APIs to just take a `Position` which is a line+character, but no URI. The goal APIs have been around so I guess we should keep them as-is for compat.

Reply:
Ok, I made this change everywhere. Can always revert if we decide otherwise.